File: dnd-movile/app/src/main/java/io/github/gasparkral/dnd/model/Inventory.kt
package io.github.gasparkral.dnd.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// ---------------------------------------------------------------------------
// Sistema de bonificadores de estadística
// ---------------------------------------------------------------------------

@Serializable
enum class BonusStat {
    @SerialName("strength")          Strength,
    @SerialName("dexterity")         Dexterity,
    @SerialName("constitution")      Constitution,
    @SerialName("intelligence")      Intelligence,
    @SerialName("wisdom")            Wisdom,
    @SerialName("charisma")          Charisma,
    @SerialName("armor_class")       ArmorClass,
    @SerialName("max_hp")            MaxHp,
    @SerialName("speed")             Speed,
    @SerialName("attack_bonus")      AttackBonus,
    @SerialName("damage_bonus")      DamageBonus,
    @SerialName("saving_throw_str")  SavingThrowStr,
    @SerialName("saving_throw_dex")  SavingThrowDex,
    @SerialName("saving_throw_con")  SavingThrowCon,
    @SerialName("saving_throw_int")  SavingThrowInt,
    @SerialName("saving_throw_wis")  SavingThrowWis,
    @SerialName("saving_throw_cha")  SavingThrowCha;

    fun label() = when (this) {
        Strength        -> "FUE"
        Dexterity       -> "DES"
        Constitution    -> "CON"
        Intelligence    -> "INT"
        Wisdom          -> "SAB"
        Charisma        -> "CAR"
        ArmorClass      -> "CA"
        MaxHp           -> "PG máx"
        Speed           -> "Veloc."
        AttackBonus     -> "Ataque"
        DamageBonus     -> "Daño"
        SavingThrowStr  -> "Sal. FUE"
        SavingThrowDex  -> "Sal. DES"
        SavingThrowCon  -> "Sal. CON"
        SavingThrowInt  -> "Sal. INT"
        SavingThrowWis  -> "Sal. SAB"
        SavingThrowCha  -> "Sal. CAR"
    }

    /** Emoji decorativo para la UI */
    fun emoji() = when (this) {
        Strength        -> "💪"
        Dexterity       -> "🎯"
        Constitution    -> "❤️"
        Intelligence    -> "📚"
        Wisdom          -> "👁️"
        Charisma        -> "✨"
        ArmorClass      -> "🛡️"
        MaxHp           -> "❤️"
        Speed           -> "💨"
        AttackBonus     -> "⚔️"
        DamageBonus     -> "🔥"
        SavingThrowStr,
        SavingThrowDex,
        SavingThrowCon,
        SavingThrowInt,
        SavingThrowWis,
        SavingThrowCha  -> "🏙️"
    }

    /** True si este stat es un atributo base (FUE/DES/CON/INT/SAB/CAR) */
    fun isBaseAttribute() = this in setOf(
        Strength, Dexterity, Constitution, Intelligence, Wisdom, Charisma
    )
}

@Serializable
enum class BonusType {
    @SerialName("item")          Item,
    @SerialName("circumstance")  Circumstance,
    @SerialName("status")        Status,
    @SerialName("untyped")       Untyped;

    /**
     * Indica si este tipo se apila con otros del mismo tipo.
     * En D&D 5.5e, Status NO se apila — solo cuenta el mayor.
     * El resto (Item, Circumstance, Untyped) sí se apilan.
     */
    fun stacks() = this != Status
}

@Serializable
data class StatBonus(
    val stat: BonusStat,
    val value: Int,
    @SerialName("bonus_type") val bonusType: BonusType = BonusType.Untyped,
    val source: String = "",
)

// ---------------------------------------------------------------------------
// Stats efectivos calculados en cliente
// ---------------------------------------------------------------------------

/**
 * Resultado de aplicar todos los bonificadores de ítems equipados
 * sobre los atributos base del personaje.
 *
 * Es la fuente de verdad para mostrar stats en Dashboard, Inventario y Combate.
 */
data class EffectiveStats(
    /** Atributos base del personaje (sin modificadores de equipo) */
    val base: AttributesDto,
    /** Suma neta de bonificadores por atributo base, respetando reglas de apilamiento */
    val attributeBonuses: Map<BonusStat, Int>,
    /** Bonus neto a CA (suma de todos los ítems equipados) */
    val armorClassBonus: Int,
    /** Bonus neto a PG máximos */
    val maxHpBonus: Int,
    /** Bonus neto a velocidad en pies */
    val speedBonus: Int,
    /** Bonus neto a ataque */
    val attackBonus: Int,
    /** Bonus neto a daño */
    val damageBonus: Int,
    /** Bonus a tiradas de salvación por atributo */
    val savingThrowBonuses: Map<BonusStat, Int>,
    /** Lista plana de todos los bonus activos con su fuente — para mostrar en tooltips */
    val activeBonuses: List<ActiveBonus>,
) {
    /** Atributo efectivo final: base + bonus de equipo */
    fun effectiveStrength()     = base.strength     + (attributeBonuses[BonusStat.Strength]     ?: 0)
    fun effectiveDexterity()    = base.dexterity    + (attributeBonuses[BonusStat.Dexterity]    ?: 0)
    fun effectiveConstitution() = base.constitution + (attributeBonuses[BonusStat.Constitution] ?: 0)
    fun effectiveIntelligence() = base.intelligence + (attributeBonuses[BonusStat.Intelligence] ?: 0)
    fun effectiveWisdom()       = base.wisdom       + (attributeBonuses[BonusStat.Wisdom]       ?: 0)
    fun effectiveCharisma()     = base.charisma     + (attributeBonuses[BonusStat.Charisma]     ?: 0)

    fun hasAnyBonus() = activeBonuses.isNotEmpty()
}

/** Bonus activo individual — para mostrar en tooltips y panel de equipo */
data class ActiveBonus(
    val stat: BonusStat,
    val value: Int,
    val bonusType: BonusType,
    val source: String,  // nombre del ítem que lo provee
)

// ---------------------------------------------------------------------------
// Cálculo de stats efectivos
// ---------------------------------------------------------------------------

/**
 * Calcula los [EffectiveStats] de un personaje aplicando todos los
 * [StatBonus] de los ítems que están equipados actualmente.
 *
 * Reglas de apilamiento (D&D 5.5e):
 * - BonusType.Item, Circumstance, Untyped: se suman todos
 * - BonusType.Status: solo cuenta el mayor valor para cada stat
 *
 * @param base      Atributos base del personaje
 * @param inventory Lista completa del inventario (se filtran los equipados)
 */
fun calculateEffectiveStats(base: AttributesDto, inventory: List<InventoryItem>): EffectiveStats {
    val equipped = inventory.filter { it.equipped && it.statBonuses.isNotEmpty() }

    // Construir lista plana de todos los bonuses activos
    val allBonuses: List<ActiveBonus> = equipped.flatMap { item ->
        item.statBonuses.map { bonus ->
            ActiveBonus(
                stat      = bonus.stat,
                value     = bonus.value,
                bonusType = bonus.bonusType,
                source    = item.name,
            )
        }
    }

    /**
     * Aplica las reglas de apilamiento para un [BonusStat] concreto.
     * Devuelve la suma neta de todos los bonuses que aplican a ese stat.
     */
    fun netBonus(stat: BonusStat): Int {
        val forStat = allBonuses.filter { it.stat == stat }
        if (forStat.isEmpty()) return 0

        // Separar por tipo y aplicar reglas
        val statusBonuses      = forStat.filter { it.bonusType == BonusType.Status }
        val stackingBonuses    = forStat.filter { it.bonusType != BonusType.Status }

        // Status: solo el mayor (positivo) o el menor (negativo, debuffs)
        val statusNet = when {
            statusBonuses.isEmpty() -> 0
            statusBonuses.any { it.value > 0 } ->
                statusBonuses.filter { it.value > 0 }.maxOf { it.value }
            else ->
                statusBonuses.minOf { it.value }  // debuff negativo
        }

        // Item / Circumstance / Untyped: se suman todos
        val stackingNet = stackingBonuses.sumOf { it.value }

        return statusNet + stackingNet
    }

    // Calcular bonuses por categoría
    val attrStats = listOf(
        BonusStat.Strength, BonusStat.Dexterity, BonusStat.Constitution,
        BonusStat.Intelligence, BonusStat.Wisdom, BonusStat.Charisma,
    )
    val savingStats = listOf(
        BonusStat.SavingThrowStr, BonusStat.SavingThrowDex, BonusStat.SavingThrowCon,
        BonusStat.SavingThrowInt, BonusStat.SavingThrowWis, BonusStat.SavingThrowCha,
    )

    val attributeBonuses  = attrStats.associateWith { netBonus(it) }.filterValues { it != 0 }
    val savingBonuses     = savingStats.associateWith { netBonus(it) }.filterValues { it != 0 }

    return EffectiveStats(
        base               = base,
        attributeBonuses   = attributeBonuses,
        armorClassBonus    = netBonus(BonusStat.ArmorClass),
        maxHpBonus         = netBonus(BonusStat.MaxHp),
        speedBonus         = netBonus(BonusStat.Speed),
        attackBonus        = netBonus(BonusStat.AttackBonus),
        damageBonus        = netBonus(BonusStat.DamageBonus),
        savingThrowBonuses = savingBonuses,
        activeBonuses      = allBonuses,
    )
}

// ---------------------------------------------------------------------------
// Extensiones de utilidad
// ---------------------------------------------------------------------------

/** Modificador D&D (score - 10) / 2, redondeando hacia abajo */
fun attributeModifier(score: Int): Int = Math.floorDiv(score - 10, 2)

/** Formatea un modificador como "+N" o "−N" */
fun formatModifier(mod: Int): String = if (mod >= 0) "+$mod" else "−${-mod}"

// ---------------------------------------------------------------------------

@Serializable
enum class ItemCategory {
    @SerialName("weapon")     Weapon,
    @SerialName("armour")     Armour,
    @SerialName("consumable") Consumable,
    @SerialName("tool")       Tool,
    @SerialName("treasure")   Treasure,
    @SerialName("accessory")  Accessory,
    @SerialName("misc")       Misc;

    fun label() = when (this) {
        Weapon    -> "Arma"
        Armour    -> "Armadura"
        Consumable -> "Consumible"
        Tool      -> "Herramienta"
        Treasure  -> "Tesoro"
        Accessory -> "Accesorio"
        Misc      -> "Misc"
    }

    fun emoji() = when (this) {
        Weapon    -> "⚔️"
        Armour    -> "🛡️"
        Consumable -> "🧪"
        Tool      -> "🔧"
        Treasure  -> "💎"
        Accessory -> "💍"
        Misc      -> "📦"
    }
}

@Serializable
data class InventoryItem(
    val id: String,
    val name: String,
    val category: ItemCategory,
    @SerialName("description") val description: String = "",
    val quantity: Int,
    val weight: Float? = null,
    val equipped: Boolean = false,
    @SerialName("accessory_type") val accessoryType: String? = null,
    @SerialName("stat_bonuses") val statBonuses: List<StatBonus> = emptyList(),
    val notes: String = "",
) {
    /** True si este ítem aporta bonificadores cuando está equipado */
    val hasStatBonuses: Boolean get() = statBonuses.isNotEmpty()

    /** True si este ítem es equipable por su categoría */
    val isEquippable: Boolean get() = category in setOf(
        ItemCategory.Weapon, ItemCategory.Armour, ItemCategory.Accessory
    )
}

@Serializable
data class Currency(
    val copper: Int = 0,
    val silver: Int = 0,
    val electrum: Int = 0,
    val gold: Int = 0,
    val platinum: Int = 0,
)

@Serializable
data class InventoryResponse(
    val items: List<InventoryItem>,
    val currency: Currency,
    @SerialName("total_weight") val totalWeight: Float,
)

@Serializable
data class AddItemRequest(
    val name: String,
    val category: ItemCategory,
    val description: String = "",
    val quantity: Int,
    val weight: Float? = null,
    @SerialName("accessory_type") val accessoryType: String? = null,
    @SerialName("stat_bonuses") val statBonuses: List<StatBonus> = emptyList(),
    val notes: String = "",
)

@Serializable
data class UpdateItemRequest(
    val quantity: Int? = null,
    val equipped: Boolean? = null,
    val notes: String? = null,
)

@Serializable
data class UpdateCurrencyRequest(
    val copper: Int? = null,
    val silver: Int? = null,
    val electrum: Int? = null,
    val gold: Int? = null,
    val platinum: Int? = null,
)

File: shared/src/api_types/inventory.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// Sistema de bonificadores de estadística
// ---------------------------------------------------------------------------

/// Estadística que puede ser modificada por un objeto equipado.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BonusStat {
    Strength,
    Dexterity,
    Constitution,
    Intelligence,
    Wisdom,
    Charisma,
    ArmorClass,
    MaxHp,
    Speed,
    AttackBonus,
    DamageBonus,
    SavingThrowStr,
    SavingThrowDex,
    SavingThrowCon,
    SavingThrowInt,
    SavingThrowWis,
    SavingThrowCha,
}

impl BonusStat {
    pub fn label(&self) -> &'static str {
        match self {
            Self::Strength         => "FUE",
            Self::Dexterity        => "DES",
            Self::Constitution     => "CON",
            Self::Intelligence     => "INT",
            Self::Wisdom           => "SAB",
            Self::Charisma         => "CAR",
            Self::ArmorClass       => "CA",
            Self::MaxHp            => "PG máx",
            Self::Speed            => "Velocidad",
            Self::AttackBonus      => "Ataque",
            Self::DamageBonus      => "Daño",
            Self::SavingThrowStr   => "Sal. FUE",
            Self::SavingThrowDex   => "Sal. DES",
            Self::SavingThrowCon   => "Sal. CON",
            Self::SavingThrowInt   => "Sal. INT",
            Self::SavingThrowWis   => "Sal. SAB",
            Self::SavingThrowCha   => "Sal. CAR",
        }
    }
}

/// Tipo de bonificador — determina las reglas de apilamiento (D&D 5.5e).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BonusType {
    /// Bonus de objeto equipado — se apilan entre sí
    Item,
    /// Bonus circunstancial — se apilan
    Circumstance,
    /// Bonus de estado — NO se apilan, solo el mayor cuenta
    Status,
    /// Sin tipo — siempre se apila
    Untyped,
}

impl BonusType {
    fn default_untyped() -> BonusType {
        BonusType::Untyped
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StatBonus {
    pub stat: BonusStat,
    pub value: i16,
    #[serde(default = "BonusType::default_untyped")]
    pub bonus_type: BonusType,
    #[serde(default)]
    pub source: String,
}

// ---------------------------------------------------------------------------
// Categorías de objeto
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ItemCategory {
    Weapon,
    Armour,
    Consumable,
    Tool,
    Treasure,
    Accessory,
    Misc,
}

impl ItemCategory {
    pub fn label(&self) -> &'static str {
        match self {
            Self::Weapon     => "Arma",
            Self::Armour     => "Armadura",
            Self::Consumable => "Consumible",
            Self::Tool       => "Herramienta",
            Self::Treasure   => "Tesoro",
            Self::Accessory  => "Accesorio",
            Self::Misc       => "Misc",
        }
    }
}

// ---------------------------------------------------------------------------
// Objeto del inventario
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InventoryItem {
    pub id: Uuid,
    pub name: String,
    pub category: ItemCategory,
    #[serde(default)]
    pub description: String,
    pub quantity: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub weight: Option<f32>,
    #[serde(default)]
    pub equipped: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accessory_type: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub stat_bonuses: Vec<StatBonus>,
    #[serde(default)]
    pub notes: String,
}

impl InventoryItem {
    pub fn new(
        name: impl Into<String>,
        category: ItemCategory,
        description: impl Into<String>,
        quantity: u32,
    ) -> Self {
        Self {
            id: Uuid::new_v4(),
            name: name.into(),
            category,
            description: description.into(),
            quantity,
            weight: None,
            equipped: false,
            accessory_type: None,
            stat_bonuses: Vec::new(),
            notes: String::new(),
        }
    }
}

// ---------------------------------------------------------------------------
// Monedas
// ---------------------------------------------------------------------------

// Bug 1 corregido: orden de campos consistente con models::inventory::Currency
// (copper → silver → electrum → gold → platinum en ambos structs).
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct Currency {
    #[serde(default)]
    pub copper: u32,
    #[serde(default)]
    pub silver: u32,
    #[serde(default)]
    pub electrum: u32,
    #[serde(default)]
    pub gold: u32,
    #[serde(default)]
    pub platinum: u32,
}

// ---------------------------------------------------------------------------
// Requests / Responses
// ---------------------------------------------------------------------------

// Bug 5 corregido en todos los request types que solo derivaban Deserialize:
// eliminados los `skip_serializing_if` que eran ruido inoperante en structs
// sin Serialize. Se añade Serialize + Clone a todos para uniformidad y para
// que puedan viajar por WebSocket / logs sin tocar shared en el futuro.

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddItemRequest {
    pub name: String,
    pub category: ItemCategory,
    #[serde(default)]
    pub description: String,
    pub quantity: u32,
    #[serde(default)]
    pub weight: Option<f32>,
    #[serde(default)]
    pub accessory_type: Option<String>,
    #[serde(default)]
    pub stat_bonuses: Vec<StatBonus>,
    #[serde(default)]
    pub notes: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateItemRequest {
    #[serde(default)]
    pub quantity: Option<u32>,
    #[serde(default)]
    pub equipped: Option<bool>,
    #[serde(default)]
    pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateCurrencyRequest {
    #[serde(default)]
    pub copper: Option<u32>,
    #[serde(default)]
    pub silver: Option<u32>,
    #[serde(default)]
    pub electrum: Option<u32>,
    #[serde(default)]
    pub gold: Option<u32>,
    #[serde(default)]
    pub platinum: Option<u32>,
}

#[derive(Debug, Serialize)]
pub struct InventoryResponse {
    pub items: Vec<InventoryItem>,
    pub currency: Currency,
    /// Peso total de todos los objetos (suma de weight * quantity)
    pub total_weight: f32,
}

impl InventoryResponse {
    pub fn from_parts(items: Vec<InventoryItem>, currency: Currency) -> Self {
        let total_weight = items
            .iter()
            .map(|i| i.weight.unwrap_or(0.0) * i.quantity as f32)
            .sum();
        Self { items, currency, total_weight }
    }
}
